sdk/eventhub/Microsoft.Azure.WebJobs.Extensions.EventHubs/src/Listeners/TargetScaleThrottler.cs

```diff
@@ -0,0 +1,147 @@
+
+        private const int ScaleUpFactor = 10;
+
+        private readonly List<EventHubTragetScalerReult> _history = new List<EventHubTragetScalerReult>();
+        private readonly ILogger _logger;
+        private object lockObject = new object();
+
+        private TargetScalerResult _lastTargetScalerResult = new TargetScalerResult();
+        private DateTime _lastScaleUpTime = DateTime.MinValue;
+
+        public TargetScaleThrottler(ILogger logger)
+        {
+            _logger = logger;
+        }
+
+        // for tests
+        public TargetScaleThrottler(DateTime lastScaleUpTime, TargetScalerResult lastTargetScalerResult, ILogger logger)
+        {
+            _lastScaleUpTime = lastScaleUpTime;
+            _lastTargetScalerResult = lastTargetScalerResult;
+            _logger = logger;
+        }
+
+        public TargetScalerResult ThrottleIfNeeded(TargetScalerResult result, EventHubsTriggerMetrics metric, DateTime executionTime)
+        {
+            try
+            {
+                lock (lockObject)
+                {
+                    // Removing expired items from history. We want to keep the history item around the threshold so adding 2 secs.
+                    var threshold = executionTime - TimeSpan.FromSeconds(ExpirationThresholdInseconds + 2);
+                    bool historyCleared = false;
+                    for (int i = _history.Count - 1; i >= 0; i--)
+                    {
+                        if (_history[i].Metrics.Timestamp < threshold)
+                        {
+                            historyCleared = true;
+                            _history.RemoveAt(i);
+                        }
+                    }
+
+                    // Adding new history item
+                    EventHubTragetScalerReult currentResult = new EventHubTragetScalerReult
+                    {
+                        TargetScalerResult = result,
+                        Metrics = metric,
+                        ChangeWorkerCount = result.TargetWorkerCount - _lastTargetScalerResult.TargetWorkerCount
+                    };
+                    _history.Add(currentResult);
+                    int changeWorkerCount = currentResult.ChangeWorkerCount;
+
+                    // Throttle scale down based on latest scale up
+                    if (changeWorkerCount < 0 && (executionTime - _lastScaleUpTime).TotalSeconds < ThrottleScaleDownIntervalInSeconds)
+                    {
+                        _logger.LogInformation($"Throttling scale down, since last scale up time was within {ThrottleScaleDownIntervalInSeconds} seconds. (LastScaleUpTime: '{_lastScaleUpTime}', LastTargetWorkerRequest: '{_lastTargetScalerResult.TargetWorkerCount}')");
+                        return _lastTargetScalerResult;
+                    }
+
+                    // If there we have enough history try to throttle based on history
+                    if ((historyCleared) && (Math.Abs(changeWorkerCount) > 0))
```

Comment:
Can you clarify why we only do this throttling if `historyCleared` is true?

Reply:
Also, since this is throttling scale up, do we need the `Math.Abs`?

Reply:
We do not want to start hitory-based-throttling within 1 minute after start monitoring. `historyCleared = true` means we have `history > 1 minute`.